### ISSUE TITLE
Model loss not improving at all

### ISSUE THREAD

**Issue by vsuriya93** [fchollet/keras#1478]
The accuracy is not at all improving.. It is predicting all the classes as same. I am working on using deep learning for image classification, in CIFAR data set. Here is the model code I am using.

model=Sequential()
model.add(Convolution2D(32,3,3,border_mode='same',input_shape=(3,32,32)))
model.add(Activation('relu'))
model.add(Convolution2D(32, 3, 3))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Convolution2D(64, 3, 3, border_mode='same'))
model.add(Activation('relu'))
model.add(Convolution2D(64, 3, 3))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(nb_classes))
model.add(Activation('softmax'))
sgd = SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='categorical_crossentropy', optimizer=sgd) # error can also be mse

I have binarized the labels, their respective shapes are
train -> (40000,3,32,32)
test -> (10000,3,32,32)

total images =50000
I have even changed the learning rate between (.1,.0001) but all the instances are predicted into the same class
What can be the possible problem.

**Comment by jliemansifry** [comment]
Could you paste the full code here for which the model won't learn? More information would help debug this issue.

**Comment by lukedeo** [comment]
Also, your labels should be of shape `(40000, nb_classes)`, not `(40000, 3, 32, 32)`.

**Comment by vsuriya93** [comment]
Here is the entire code for the model

from __future__ import print_function
from keras.preprocessing.image import ImageDataGenerator
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation, Flatten
from keras.layers.convolutional import Convolution2D, MaxPooling2D
from keras.optimizers import SGD, Adadelta, Adagrad
from keras.utils import np_utils, generic_utils
from six.moves import range
import pandas as pd
import numpy as np

batch_size = 32
nb_classes = 10
nb_epoch = 200
img_rows, img_cols = 32, 32
img_channels = 3


df=pd.read_csv('train.csv',header=None) #contains the training set, shape (50000,3072)
label=pd.read_csv('trainLabels.csv') #contains labels (50000,1) ( as text, like frog, plane etc)

# replacing labels by int values (between 0-9)
count=0
mapper={}
for x in label.label.unique():
    mapper[x]=count
    count=count+1 
label.label.replace(mapper,inplace=True)
label=np.array(label.label)  # label has the class labels 0-9

# spliting into test and train set 

from sklearn.cross_validation import train_test_split
train,test,train_label,test_label=train_test_split(df,label,test_size=.20,random_state=33)

# preprocessing the labels

y_train=np_utils.to_categorical(train_label,nb_classes)
y_test=np_utils.to_categorical(test_label,nb_classes)
train=train.reshape(-1,3,32,32)
test=test.reshape(-1,3,32,32)
train=train.astype('float32')
test=test.astype('float32')
train=train/255
test=test/255

# building the model

model=Sequential()

model.add(Convolution2D(32,3,3,border_mode='same',input_shape=(3,32,32)))
model.add(Activation('relu'))
model.add(Convolution2D(32, 3, 3))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

model.add(Convolution2D(64, 3, 3, border_mode='same'))
model.add(Activation('relu'))
model.add(Convolution2D(64, 3, 3))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

model.add(Flatten())

model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(nb_classes))
model.add(Activation('softmax'))

sgd = SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='categorical_crossentropy', optimizer=sgd)
output=model.predict_classes(test)

**Comment by jliemansifry** [comment]
I think you're forgetting to actually fit the model. After compiling, try something along the lines of:

```python
model.fit(train, y_train, nb_epoch=16, batch_size=32, verbose=1, show_accuracy=True)
```

After fitting, the model should be predicting different classes.

**Comment by vsuriya93** [comment]
I had actually used the fit function, but still no improvement what so ever. I paste the screen shot below
![screenshot from 2016-01-20 20 46 50](https://cloud.githubusercontent.com/assets/7788805/12452946/f90ca222-bfb6-11e5-8aed-da1b1b4cda8d.png)

All the elements are classified into class 2 !!

**Comment by jliemansifry** [comment]
Ok. 

I'd try using a different optimizer (adadelta) and messing around with some of your parameters in your net (particularly dropout) to see if you can get it learning.

**Comment by tboquet** [comment]
You could try a really small learning rate (`1e-7`) just to check that it's a learning problem. With this rate, you should observe a decreasing loss. Another point, what are the relative frequencies of each classes? Did you checked descriptive statistics of your inputs or to visualize it along with your classes?

**Comment by vsuriya93** [comment]
Still no improvement what so ever.. I am pasting the screen shot below

![screenshot from 2016-01-21 11 58 39](https://cloud.githubusercontent.com/assets/7788805/12473244/7fce3b2a-c036-11e5-93e2-81a3a4335fdd.png)

**Comment by fchollet** [comment]
For a working CIFAR10 example, you should check out the one in the Keras
examples folder.

**Comment by vsuriya93** [comment]
I implemented from the example given in CIFAR10 example folder, only the data set I obtained was from Kaggle. Can there be a problem with the data set then? Should I try the one given in CIFAR page?

**Comment by tboquet** [comment]
Are you able to run the cifar10 exemple? If so, start with the exact same script an adapt the inputs/outputs so it can be fed to the model. If it's not working, you should visualize both inputs and outputs to try to understand what is going on.

**Comment by vsuriya93** [comment]
Yes, I ran the CIFAR10 example on 10000 examples, and it gave 90% accuracy.. Thank you, but still I am wondering why it did not work for the dataset provided in kaggle ??

Link for the CIFAR data as given in Kaggle:  [https://www.kaggle.com/c/cifar-10]

Any thoughts on that ??

**Comment by r3v1** [comment]
Any solutions? I am in the same situation as @vsuriya93.  Look at this:

I have created the next model using `Keras` module, compiled and fitted with custom images:

```
from keras import Sequential
from keras.layers import Conv2D, MaxPooling2D, Activation, Flatten, Dense, BatchNormalization, ZeroPadding2D, Dropout
from keras.regularizers import l2
import keras.optimizers as Optimizer

n_classes = 3

# Initialize model
model = Sequential()

# 2 sets of CRP (Convolution, RELU, Pooling)
model.add(Conv2D(20, (5, 5), padding="same", input_shape=(*img_shape, 1), kernel_regularizer=l2(0)))
model.add(Activation("relu"))
model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))

model.add(Conv2D(50, (5, 5), padding="same", kernel_regularizer=l2(0)))
model.add(Activation("relu"))
model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))

# Fully connected layers (w/ RELU)
model.add(Flatten())
model.add(Dense(500, kernel_regularizer=l2(0)))
model.add(Activation("relu"))

# Softmax (for classification)
model.add(Dense(n_classes, kernel_regularizer=l2(0)))
model.add(Activation("softmax"))

model.compile(loss='categorical_crossentropy', optimizer=Optimizer.Adam(lr=1e-7), metrics=['accuracy'])

model.fit(X_train, y_train, batch_size=128, epochs=5, validation_data=(X_test, y_test))
```
and receiving the next output:

> Train on 477 samples, validate on 235 samples
> Epoch 1/5
> 477/477 [==============================] - 1s 2ms/step - loss: 10.5764 - acc: 0.3438 - val_loss: 11.3170 - val_acc: 0.2979
> Epoch 2/5
> 477/477 [==============================] - 0s 657us/step - loss: 10.5764 - acc: 0.3438 - val_loss: 11.3170 - val_acc: 0.2979
> Epoch 3/5
> 477/477 [==============================] - 0s 662us/step - loss: 10.5764 - acc: 0.3438 - val_loss: 11.3170 - val_acc: 0.2979
> Epoch 4/5
> 477/477 [==============================] - 0s 685us/step - loss: 10.5764 - acc: 0.3438 - val_loss: 11.3170 - val_acc: 0.2979
> Epoch 5/5
> 477/477 [==============================] - 0s 674us/step - loss: 10.5764 - acc: 0.3438 - val_loss: 11.3170 - val_acc: 0.2979

Well. Know importing `tensorflow.python.keras` instead of `keras`:

```
from tensorflow.python.keras import Sequential
from tensorflow.python.keras.layers import Conv2D, MaxPooling2D, Activation, Flatten, Dense, BatchNormalization, \
    ZeroPadding2D, Dropout
from tensorflow.python.keras.regularizers import l2
import tensorflow.python.keras.optimizers as Optimizer

n_classes = 3

# Initialize model
model = Sequential()

# 2 sets of CRP (Convolution, RELU, Pooling)
model.add(Conv2D(20, (5, 5), padding="same", input_shape=(*img_shape, 1), kernel_regularizer=l2(0)))
model.add(Activation("relu"))
model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))

model.add(Conv2D(50, (5, 5), padding="same", kernel_regularizer=l2(0)))
model.add(Activation("relu"))
model.add(MaxPooling2D(pool_size=(2, 2), strides=(2, 2)))

# Fully connected layers (w/ RELU)
model.add(Flatten())
model.add(Dense(500, kernel_regularizer=l2(0)))
model.add(Activation("relu"))

# Softmax (for classification)
model.add(Dense(n_classes, kernel_regularizer=l2(0)))
model.add(Activation("softmax"))

model.compile(loss='categorical_crossentropy', optimizer=Optimizer.Adam(lr=1e-7), metrics=['accuracy'])

model.fit(X_train, y_train, batch_size=128, epochs=5, validation_data=(X_test, y_test))
```

getting the next output:

> Train on 477 samples, validate on 235 samples
> Epoch 1/5
> 477/477 [==============================] - 1s 1ms/sample - loss: 33.8133 - acc: 0.3396 - val_loss: 33.9784 - val_acc: 0.3489
> Epoch 2/5
> 477/477 [==============================] - 0s 739us/sample - loss: 33.5633 - acc: 0.3396 - val_loss: 33.7300 - val_acc: 0.3489
> Epoch 3/5
> 477/477 [==============================] - 0s 666us/sample - loss: 33.3153 - acc: 0.3396 - val_loss: 33.4805 - val_acc: 0.3489
> Epoch 4/5
> 477/477 [==============================] - 0s 645us/sample - loss: 33.0616 - acc: 0.3396 - val_loss: 33.2312 - val_acc: 0.3489
> Epoch 5/5
> 477/477 [==============================] - 0s 635us/sample - loss: 32.8133 - acc: 0.3396 - val_loss: 32.9808 - val_acc: 0.3489

It reduces loss! I don't know why, but importing `tensorflow.python.keras` instead of 'keras' works as expected. Any reason for that?
